components/supervisor/pkg/serverApi/publicapi.go

```diff
@@ -0,0 +1,285 @@
+	configCat "github.com/gitpod-io/gitpod/common-go/experiments"
+	"github.com/gitpod-io/gitpod/common-go/log"
+	v1 "github.com/gitpod-io/gitpod/components/public-api/go/experimental/v1"
+	gitpod "github.com/gitpod-io/gitpod/gitpod-protocol"
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+type ServerAPIInterface interface {
+	GetWorkspace(ctx context.Context, id string) (res *gitpod.WorkspaceInfo, err error)
+	GetToken(ctx context.Context, query *gitpod.GetTokenSearchOptions) (res *gitpod.Token, err error)
+	OpenPort(ctx context.Context, workspaceID string, port *gitpod.WorkspaceInstancePort) (res *gitpod.WorkspaceInstancePort, err error)
+	InstanceUpdates(ctx context.Context, instanceID string, workspaceID string) (<-chan *gitpod.WorkspaceInstance, error)
+
+	// Remove this and use segment client directly
+	TrackEvent(ctx context.Context, event *gitpod.RemoteTrackMessage) (err error)
+}
+
+const (
+	// KindGitpod marks tokens that provide access to the Gitpod server API.
+	KindGitpod = "gitpod"
+)
+
+type ServerApiService struct {
+	// cfg *Config
+	instanceID  string
+	workspaceID string
+
+	// gitpodService server API
+	gitpodService gitpod.APIInterface
+	// publicAPIConn public API publicAPIConn
+	publicAPIConn *grpc.ClientConn
+}
+
+var _ ServerAPIInterface = (*ServerApiService)(nil)
+
+func usePublicAPI(ctx context.Context, ownerID string, experiments configCat.Client) bool {
+	return ownerID != "" && experiments.GetBoolValue(ctx, "supervisor_experimental_publicapi", false, configCat.Attributes{
+		UserID: ownerID,
+	})
+}
+
+func NewServerApiService(endpoint string, host string, instanceID string, workspaceID string, version string, tknsrv api.TokenServiceServer, experiments configCat.Client) *ServerApiService {
+	tknres, err := tknsrv.GetToken(context.Background(), &api.GetTokenRequest{
+		Kind: KindGitpod,
+		Host: host,
+		Scope: []string{
+			"function:getToken",
+			"function:openPort",
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("cannot get token for Gitpod API")
+		return nil
+	}
+
+	gitpodService, err := gitpod.ConnectToServer(endpoint, gitpod.ConnectToServerOpts{
+		Token: tknres.Token,
+		Log:   log.Log,
+		ExtraHeaders: map[string]string{
+			"User-Agent":              "gitpod/supervisor",
+			"X-Workspace-Instance-Id": instanceID,
+			"X-Client-Version":        version,
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("cannot connect to Gitpod API")
+		return nil
+	}
+
+	resp := &ServerApiService{
+		gitpodService: gitpodService,
+		// conn:          conn,
+
+		instanceID:  instanceID,
+		workspaceID: workspaceID,
+	}
+
+	wsInfo, err := gitpodService.GetWorkspace(context.TODO(), workspaceID)
+	if err != nil {
+		log.WithError(err).Error("cannot get workspace info")
+		return resp
+	}
+
+	if !usePublicAPI(context.TODO(), wsInfo.Workspace.OwnerID, experiments) {
```

Comment:
we should evaluate it for each request to be able to shift users quickly, not only workspace restart

Reply:
That sounds like an edgy case (and complicate things) that would imply the API is really flaky and would not really make sense to ship to end users in the first place, I'd expect to ship this to whole gitpod team first and after we are sure it's stable then start shipping to a handful of end users.

BTW, we have public api in gitpod-desktop for over a month now and it's still only enable for gitpod team, I'd say we should experiment there with end users first before doing it in supervisor which is a really critical component

Reply:
:wave: @akosyakov we use [loadgen](https://github.com/gitpod-io/gitpod/tree/main/dev/loadgen) to do node, cluster, and autoscaler testing. We generally do load testing in [ephemeral clusters](https://www.notion.so/How-to-create-delete-ephemeral-workspace-cluster-2e636dae17d148a5a7de5afd54566479), which you can create too.

What sort of load test are you hoping to run? For example, how quickly workspaces start, max # of running workspaces supported, how a new feature impacts runtime performance? I ask because we can apply feature flags as part of the test, but, it is not using configcat (we don't start workspaces through server with loadgen, we do direct via ws-manager).